pkgs/dart_model/benchmark/json_buffer.dart

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:typed_data';
+import 'package:benchmark_harness/benchmark_harness.dart';
+import 'package:dart_model/src/json_buffer.dart';
+
+Uint8List? serialized;
+JsonBuffer? deserialized;
+const mapSize = 1000000;
+final mapKeys = List.generate(mapSize, (i) => i.toString());
+
+void main() {
+  JsonBufferSerializeBenchmark().report();
+  // See note below, this is not measuring anything right now.
+  JsonBufferDeserializeBenchmark().report();
+  if (deserialized!.asMap.keys.length != mapSize) {
+    throw StateError('Benchmark invalid, map was not of expected size');
+  }
+}
+
+class JsonBufferSerializeBenchmark extends BenchmarkBase {
+  JsonBufferSerializeBenchmark() : super('JsonBufferSerialize');
+
+  @override
+  void run() {
+    serialized = JsonBuffer(LazyMap(mapKeys, (key) => key)).serialize();
+  }
+}
+
+class JsonBufferDeserializeBenchmark extends BenchmarkBase {
+  JsonBufferDeserializeBenchmark() : super('JsonBufferDeserialize');
+
+  @override
+  void run() {
+    // TODO: This is actually a no-op, so the benchmark is kind of pointless.
+    // Should we read all the keys/values?
```

Comment:
Yeah it's a fun one; when I was playing with this I tried two things: deep copy out to a normal map, and "process", which was just computing a hash code over everything.

https://github.com/dart-lang/language/blob/5527a8f2825f1b9fa69d7efce9ba6102bfa7aa14/working/macros/dart_model/testing/json_benchmark/lib/json_benchmark.dart#L69

I think "process" probably made the most sense of the two, although it's still only a rough approximation to real use.

Reply:
Yeah I also was realizing that for the `serialize` benchmark its really the creation, not the serialization which takes up the vast majority of the time. Which is by design of course. That benchmark makes more sense though still.